File: arrays_hashings/containsDuplicate.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {boolean}
     */
    hasDuplicate(nums) {
        // // brute force O(n^2)
        // for (let i = 0; i < nums.length; i++) {
        //     for (let j = 0; j < nums.length; j++) {
        //         if (j !== i) {
        //             if (nums[i] === nums[j])
        //                 return true
        //         }
        //     }
        //     const element = array[i];
        // }
        // return false

        // O(n)
        // one way to do it
        // let theMap = {}
        // for (let i = 0; i < nums.length; i++) {
        //     if (!theMap[nums[i]]) {
        //         theMap[nums[i]] = 1
        //     } else {
        //         theMap[nums[i]] += 1
        //     }
        // }

        // for (const key in theMap) {
        //     if (theMap[key] > 1) {
        //         return true
        //     }
        // }
        // return false

        // another way
        let theMap = {}
        for (let i = 0; i < nums.length; i++) {
            if (!theMap[nums[i]]) {
                theMap[nums[i]] = 1
            } else {
                return true
            }
        }
        return false
    }
}


// const solution = new Solution()
// // const checkDuplicate = solution.hasDuplicate([1, 2, 3, 4, 5, 1, 2, 1])
// const checkDuplicate = solution.hasDuplicate([1, 2, 3, 4, 5])
// console.log(checkDuplicate)

// const set = new Set()

// set.add(1)
// set.add(2)
// console.log(set)

File: arrays_hashings/contains_duplicate.py
from typing import List

class Solution:
    def hasDuplicate(self, nums: List[int]) -> bool:
        # one way to do it O(n)
        # hash_set = set()
        # for num in nums:
        #     if num in hash_set:
        #         return True
        #     hash_set.add(num)
        # return False
    
        # one way to do it O(n)
        return len(nums) != len(set(nums))
    
def main():
    solution = Solution()
    input = [1,2,3,4,1]

    check_duplicate = solution.hasDuplicate(input)
    print(check_duplicate)

if __name__ == "__main__":
    main()
